### PR TITLE
List instances in GCE from all zones

### PR DESCRIPTION
**Pull request by cfconrad**
Like on EC2, GCE has multiple regions with multiple zones. So we need to iterate over each zone and list the instances.

### CODE REVIEW

File: webui/ocw/lib/EC2db.py [hunk 1]
```diff
@@ -35,8 +35,8 @@ def _instance_to_json(i):
 @transaction.atomic
 def sync_instances_db(region, instances):
     o = Instance.objects
-    o = o.filter(region=region, provider=ProviderChoice.EC2, state=StateChoice.ACTIVE)
-    o = o.update(state=StateChoice.UNK, active=False)
+    o = o.filter(region=region, provider=ProviderChoice.EC2)
```
Comment (asmorodskyi) [inline webui/ocw/lib/EC2db.py]: how this relates to GCE ? if not can you please explain what is happening here ?

File: webui/ocw/lib/azure.py [hunk 2]
```diff
@@ -89,8 +89,8 @@ def _instance_to_json(i):
 @transaction.atomic
 def sync_instances_db(instances):
     o = Instance.objects
-    o = o.filter(provider=ProviderChoice.AZURE, state=StateChoice.ACTIVE)
-    o = o.update(active=False, state=StateChoice.UNK)
+    o = o.filter(provider=ProviderChoice.AZURE)
```
Comment (asmorodskyi) [inline webui/ocw/lib/azure.py]: same questions , see above

File: webui/ocw/lib/EC2db.py (hunk 1 above)
Comment (cfconrad) [reply webui/ocw/lib/EC2db.py]: This is really not directly related to GCE, but I figured that out during GCE extending.
The thing is that the synchronization between local DB and the CSP status was a bit obfuscated.

It works now like:
  * Instance.active is used to identify if such entry exists on CSP
  * Instance.state is used to reflect the latest "local" state, e.g. if delete button was pressed.

Comment (asmorodskyi) [reply webui/ocw/lib/EC2db.py]: can you please add this to field description ?

Comment (asmorodskyi) [reply webui/ocw/lib/EC2db.py]: as comments where you declare Instance ( db.py )

File: webui/ocw/lib/EC2db.py [hunk 3]
```diff
@@ -35,8 +35,8 @@ def _instance_to_json(i):
...
+    o = o.update(active=False)
```
Comment (asmorodskyi) [inline webui/ocw/lib/EC2db.py]: and why it's getting active=False here ?

File: webui/ocw/lib/db.py [hunk 4]
```diff
@@ -21,7 +21,7 @@ def update_or_create_instance(provider, instance_id, region, csp_info, active=Tr
         o.age = t_now - o.first_seen
         o.active = active
         o.csp_info = json.dumps(csp_info, ensure_ascii=False)
-        if o.state == StateChoice.UNK:
+        if o.state != StateChoice.DELETING:
```
Comment (asmorodskyi) [inline webui/ocw/lib/db.py]: and if it's DELETED ? 
and if it is DELETING why we saving the state ?

File: webui/ocw/lib/gce.py [hunk 5]
```diff
@@ -30,35 +30,75 @@ def compute_client(self):
...
 
-    def list_instances(self, zone='europe-west1-b'):
-        i = self.compute_client().instances().list(project=self.__project, zone=zone).execute()
-        return i['items'] if 'items' in i else []
+    def list_instances(self, zone):
+        result = []
+        request = self.compute_client().instances().list(project=self.__project, zone=zone)
+        while request is not None:
+            response = request.execute()
+            if 'items' in response:
```
Comment (asmorodskyi) [inline webui/ocw/lib/gce.py]: and if there is no 'items' in response ? maybe we need it least log this ?

File: webui/ocw/lib/gce.py [hunk 6]
```diff
@@ -30,35 +30,75 @@ def compute_client(self):
             self.__compute_clinet = googleapiclient.discovery.build('compute', 'v1', credentials=credentials)
         return self.__compute_clinet
 
-    def list_instances(self, zone='europe-west1-b'):
-        i = self.compute_client().instances().list(project=self.__project, zone=zone).execute()
-        return i['items'] if 'items' in i else []
+    def list_instances(self, zone):
```
Comment (asmorodskyi) [inline webui/ocw/lib/gce.py]: what method is doing looks strange and requires some explanations ( even for yourself are sure that you won't  forget what's happening here in 6 months ? )

File: webui/ocw/lib/gce.py (hunk 5 above)
Comment (cfconrad) [reply webui/ocw/lib/gce.py]: Don't know. Thins only means, that there  is no instance running at CSP. Which could happen.

File: webui/ocw/lib/db.py (hunk 4 above)
Comment (cfconrad) [reply webui/ocw/lib/db.py]: Add a  warning, if we change from DELETED to something else

**Comment by asmorodskyi** [comment]
LGTM , is it ready to merge ?

**Pull request merged**